File: fe_coffee_manager/src/types/payroll.ts
// ========== Enums ==========

export type PayrollStatus = 'DRAFT' | 'REVIEW' | 'APPROVED' | 'PAID';

export type BonusType = 'PERFORMANCE' | 'ATTENDANCE' | 'SPECIAL' | 'HOLIDAY' | 'OTHER';
export type BonusStatus = 'PENDING' | 'APPROVED' | 'REJECTED';

export type PenaltyType = 'NO_SHOW' | 'LATE' | 'EARLY_LEAVE' | 'MISTAKE' | 'VIOLATION' | 'OTHER';
export type PenaltyStatus = 'PENDING' | 'APPROVED' | 'REJECTED';

export type AllowanceType = 'TRANSPORT' | 'MEAL' | 'PHONE' | 'HOUSING' | 'OTHER';
export type AllowanceStatus = 'ACTIVE' | 'INACTIVE';

// ========== Payroll Types ==========

export interface Payroll {
  payrollId: number;
  userId: number;
  userRole: string;
  branchId: number;
  period: string; // Format: YYYY-MM
  userName?: string;
  branchName?: string;
  baseSalary: number;
  baseSalarySnapshot: number;
  hourlyRateSnapshot: number;
  insuranceSalarySnapshot: number;
  overtimeHours: number;
  overtimePay: number;
  totalAllowances: number;
  totalBonuses: number;
  totalPenalties: number;
  grossSalary: number;
  amountInsurances: number;
  amountTax: number;
  amountAdvances: number;
  totalDeductions: number;
  netSalary: number;
  status: PayrollStatus;
  createdBy: number;
  approvedBy: number | null;
  approvedAt: string | null;
  paidAt: string | null;
  notes: string | null;
  createAt: string;
  updateAt: string;
}

export interface PayrollCalculationRequest {
  userId: number;
  period: string; // Format: YYYY-MM
  notes?: string;
}

export interface BatchCalculateRequest {
  userIds: number[];
  period: string; // Format: YYYY-MM
}

export interface BatchApproveRequest {
  payrollIds: number[];
}

export interface PayrollFilters {
  userId?: number;
  branchId?: number;
  period?: string;
  status?: PayrollStatus;
}

// ========== Bonus Types ==========

export interface Bonus {
  bonusId: number;
  userId: number;
  branchId: number;
  period: string; // Format: YYYY-MM
  bonusType: BonusType;
  amount: number;
  description: string;
  status: BonusStatus;
  sourceTemplateId: number | null;
  shiftId?: number | null;
  createdBy: number;
  approvedBy: number | null;
  rejectedBy: number | null;
  rejectionReason: string | null;
  approvedAt: string | null;
  rejectedAt: string | null;
  createAt: string;
  updateAt: string;
}

export interface BonusCreationRequest {
  userId: number;
  period: string; // Format: YYYY-MM
  bonusType: BonusType;
  amount: number;
  description: string;
  shiftId?: number | null;
}

export interface BonusApplyTemplateRequest {
  userId: number;
  period: string; // Format: YYYY-MM
  templateId: number;
  overrideAmount?: number;
  overrideDescription?: string;
}

export interface BonusFilters {
  userId?: number;
  branchId?: number;
  period?: string;
  status?: BonusStatus;
}

// ========== Penalty Types ==========

export interface Penalty {
  penaltyId: number;
  userId: number;
  branchId: number;
  period: string; // Format: YYYY-MM
  penaltyType: PenaltyType;
  amount: number;
  description: string;
  status: PenaltyStatus;
  sourceTemplateId: number | null;
  incidentDate?: string | null; // Format: YYYY-MM-DD
  shiftId?: number | null; // Tham chiếu đến shift
  reasonCode?: string | null; // Mã lý do
  createdBy: number;
  approvedBy: number | null;
  rejectedBy: number | null;
  rejectionReason: string | null;
  approvedAt: string | null;
  rejectedAt: string | null;
  createAt: string;
  updateAt: string;
}

export interface PenaltyCreationRequest {
  userId: number;
  period: string; // Format: YYYY-MM
  penaltyType: PenaltyType;
  amount: number;
  description: string;
  incidentDate?: string | null; // Format: YYYY-MM-DD
  shiftId?: number | null; // Tham chiếu đến shift
  reasonCode?: string | null; // Mã lý do
}

export interface PenaltyApplyTemplateRequest {
  userId: number;
  period: string; // Format: YYYY-MM
  templateId: number;
  overrideAmount?: number;
  overrideDescription?: string;
}

export interface PenaltyFilters {
  userId?: number;
  branchId?: number;
  period?: string;
  status?: PenaltyStatus;
}

// ========== Allowance Types ==========

export interface Allowance {
  allowanceId: number;
  userId: number;
  branchId: number;
  period: string; // Format: YYYY-MM
  allowanceType: AllowanceType;
  amount: number;
  description: string;
  status: AllowanceStatus;
  sourceTemplateId: number | null;
  createdBy: number;
  createAt: string;
  updateAt: string;
}

export interface AllowanceCreationRequest {
  userId: number;
  period: string; // Format: YYYY-MM
  allowanceType: AllowanceType;
  amount: number;
  description: string;
}

export interface AllowanceApplyTemplateRequest {
  userId: number;
  period: string; // Format: YYYY-MM
  templateId: number;
  overrideAmount?: number;
  overrideDescription?: string;
}

export interface AllowanceFilters {
  userId?: number;
  branchId?: number;
  period?: string;
  status?: AllowanceStatus;
}

// ========== Template Types ==========

export interface AllowanceTemplate {
  templateId: number;
  branchId: number | null; // NULL = SYSTEM scope
  name: string;
  allowanceType: AllowanceType;
  amount: number;
  description: string;
  isActive: boolean;
  createdBy: number;
  createAt: string;
  updateAt: string;
  usageCount?: number; // Số lần template đã được sử dụng (optional)
}

export interface BonusTemplate {
  templateId: number;
  branchId: number | null; // NULL = SYSTEM scope
  name: string;
  bonusType: BonusType;
  amount: number;
  description: string;
  isActive: boolean;
  createdBy: number;
  createAt: string;
  updateAt: string;
  usageCount?: number; // Số lần template đã được sử dụng (optional)
}

export interface PenaltyConfig {
  configId: number;
  branchId: number | null; // NULL = SYSTEM scope
  name: string;
  penaltyType: PenaltyType;
  amount: number;
  description: string;
  isActive: boolean;
  createdBy: number;
  createAt: string;
  updateAt: string;
  usageCount?: number; // Số lần config đã được sử dụng (optional)
}

export interface AllowanceTemplateCreationRequest {
  branchId?: number | null; // NULL = SYSTEM scope
  name: string;
  allowanceType: AllowanceType;
  amount: number;
  description?: string;
  isActive?: boolean;
}

export interface AllowanceTemplateUpdateRequest {
  name?: string;
  allowanceType?: AllowanceType;
  amount?: number;
  description?: string;
  isActive?: boolean;
}

export interface BonusTemplateCreationRequest {
  branchId?: number | null; // NULL = SYSTEM scope
  name: string;
  bonusType: BonusType;
  amount: number;
  description?: string;
  isActive?: boolean;
}

export interface BonusTemplateUpdateRequest {
  name?: string;
  bonusType?: BonusType;
  amount?: number;
  description?: string;
  isActive?: boolean;
}

export interface PenaltyConfigCreationRequest {
  branchId?: number | null; // NULL = SYSTEM scope
  name: string;
  penaltyType: PenaltyType;
  amount: number;
  description?: string;
  isActive?: boolean;
}

export interface PenaltyConfigUpdateRequest {
  name?: string;
  penaltyType?: PenaltyType;
  amount?: number;
  description?: string;
  isActive?: boolean;
}

export interface TemplateFilters {
  branchId?: number;
  isActive?: boolean;
}

// ========== Helper Types ==========

export type PayrollResponse = Payroll;
export type BonusResponse = Bonus;
export type PenaltyResponse = Penalty;
export type AllowanceResponse = Allowance;
export type PayrollTemplate = AllowanceTemplate | BonusTemplate | PenaltyConfig;

File: profile-service/src/main/java/com/service/profile/entity/Payroll.java
package com.service.profile.entity;

import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.FieldDefaults;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE)
@Entity
@Table(name = "payrolls")
public class Payroll {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "payroll_id")
    Integer payrollId;

    @Column(name = "user_id", nullable = false)
    Integer userId;

    @Enumerated(EnumType.STRING)
    @Column(name = "user_role", nullable = false, length = 20)
    UserRole userRole;

    @Column(name = "branch_id", nullable = false)
    Integer branchId;

    @Column(name = "period", nullable = false, length = 7)
    String period; // Format: YYYY-MM

    // Thành phần lương (Snapshot - lưu giá trị tại thời điểm tính)
    @Column(name = "base_salary", nullable = false, precision = 12, scale = 2)
    BigDecimal baseSalary;

    @Column(name = "base_salary_snapshot", precision = 12, scale = 2)
    BigDecimal baseSalarySnapshot;

    @Column(name = "hourly_rate_snapshot", precision = 12, scale = 2)
    BigDecimal hourlyRateSnapshot;

    @Column(name = "insurance_salary_snapshot", precision = 12, scale = 2)
    BigDecimal insuranceSalarySnapshot;

    // Overtime
    @Column(name = "overtime_hours", precision = 4, scale = 2)
    BigDecimal overtimeHours;

    @Column(name = "overtime_pay", precision = 12, scale = 2)
    BigDecimal overtimePay;

    // Các khoản cộng
    @Column(name = "total_allowances", precision = 12, scale = 2)
    BigDecimal totalAllowances;

    @Column(name = "total_bonuses", precision = 12, scale = 2)
    BigDecimal totalBonuses;

    // Các khoản trừ
    @Column(name = "total_penalties", precision = 12, scale = 2)
    BigDecimal totalPenalties;

    // Tổng lương
    @Column(name = "gross_salary", nullable = false, precision = 12, scale = 2)
    BigDecimal grossSalary;

    // Khấu trừ (tách rõ từng khoản)
    @Column(name = "amount_insurances", precision = 15, scale = 2)
    BigDecimal amountInsurances;

    @Column(name = "amount_tax", precision = 15, scale = 2)
    BigDecimal amountTax;

    @Column(name = "amount_advances", precision = 15, scale = 2)
    BigDecimal amountAdvances;

    @Column(name = "total_deductions", precision = 15, scale = 2)
    BigDecimal totalDeductions;

    // Lương thực nhận
    @Column(name = "net_salary", nullable = false, precision = 15, scale = 2)
    BigDecimal netSalary;

    // Trạng thái và quy trình
    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 20)
    PayrollStatus status;

    @Column(name = "created_by", nullable = false)
    Integer createdBy;

    @Column(name = "approved_by")
    Integer approvedBy;

    @Column(name = "approved_at")
    LocalDateTime approvedAt;

    @Column(name = "paid_at")
    LocalDateTime paidAt;

    // Ghi chú
    @Column(name = "notes", columnDefinition = "TEXT")
    String notes;

    // Timestamps
    @Column(name = "create_at", nullable = false, updatable = false,
            columnDefinition = "DATETIME DEFAULT CURRENT_TIMESTAMP")
    LocalDateTime createAt;

    @Column(name = "update_at", nullable = false,
            columnDefinition = "DATETIME DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP")
    LocalDateTime updateAt;

    public enum UserRole {
        STAFF, MANAGER
    }

    public enum PayrollStatus {
        DRAFT, REVIEW, APPROVED, PAID
    }
}

